rainier-core/src/main/scala/com/stripe/rainier/sampler/Sampler.scala

```diff
@@ -1,13 +1,85 @@
 package com.stripe.rainier.sampler
 
 import com.stripe.rainier.compute._
+import scala.annotation.tailrec
 
 trait Sampler {
   def sample(density: Real, warmupIterations: Int)(
       implicit rng: RNG): Stream[Sample]
 }
 
 final case class Sample(accepted: Boolean, parameters: Array[Double])
+final case class Diagnostics(rHat: Double, effectiveSampleSize: Double)
+
+object Diagnostics {
+  def apply(traces: Seq[Array[Double]]): Diagnostics = {
+    val m = traces.size.toDouble
+    val n = traces.head.size.toDouble
+    val (rHat, v) = rHatAndV(traces, n, m)
+    val ac = autocorrelation(traces, n, m, v, 1, 0.0)
+    val ess = n * m / (1 + (2 * ac))
+    Diagnostics(rHat, ess)
+  }
+
+  def rHatAndV(traces: Seq[Array[Double]],
```

Comment:
might be helpful to reference the stan manual sections in comments here (or at the object level)

Reply:
==